### ISSUE TITLE
[RFC] Multi-Turn Visual Reflection SFT for BAGEL in verl-omni

### ISSUE THREAD

**Issue by BorisYang326** [verl-project/verl-omni#295]
### Feature request

This RFC proposes an offline multi-turn visual-reflection SFT path for the repository-native BAGEL implementation in
verl-omni:

```text
prompt -> generate image -> reflect -> stop
                                   `-> continue -> edit -> reflect -> ...
```

The MVP trains three atomic SFT primitives using homogeneous micro-batches:

- `t2i`: prompt to a target image;
- `reflect`: prompt and current image to a structured `Reflection` / `Action` / `Edit` response;
- `edit`: prompt, current image, and delta Edit instruction to the next image.

This is a BAGEL-specific proposal. The architecture source of truth remains the existing verl-omni implementation,
especially:

- [`BagelForTraining`](https://github.com/verl-project/verl-omni/blob/main/verl_omni/pipelines/bagel_flow_grpo/bagel_model.py);
- [`BagelDiffusion`](https://github.com/verl-project/verl-omni/blob/main/verl_omni/pipelines/bagel_flow_grpo/diffusers_training_adapter.py);
- the current BAGEL checkpoint mapping, FSDP, and FlowGRPO integration; and
- `NonDiffusersModelBase`.

Public datasets and checkpoints provide data and tensors to repository-owned loaders. They do not replace verl-omni's
model structure, forward semantics, MoT routing, attention, conditioning, or loss contracts. This RFC does not import or
start from a separate standalone BAGEL implementation.

### Public datasets and data construction

The initial recipe uses exactly three public sources:

| Public dataset | Role | Construction |
| --- | --- | --- |
| [Fr0zencr4nE/UniCoT-Self-Reflection-6K](https://huggingface.co/datasets/Fr0zencr4nE/UniCoT-Self-Reflection-6K) | Native multi-turn reflection data | Directly parse its image states, evaluations, edits, and terminal decisions. |
| [Yejy53/Echo-4o-Image](https://huggingface.co/datasets/Yejy53/Echo-4o-Image), `Instruction-Following-Image` | Synthesized 0/1-turn data | Generate a repo-native BAGEL draft from each public T2I pair and use a VLM to construct and verify a zero- or one-edit trajectory. |
| [brivangl/midjourney-v6-llava](https://huggingface.co/datasets/brivangl/midjourney-v6-llava) | Synthesized k-turn data | Use only the public prompt field and iteratively apply repo-native BAGEL generation, VLM reflection, and an edit model. |

Every data release must pin the source revision and record dataset, model, prompt-template, filtering, seed, and image-hash
provenance.

The Midjourney dataset card includes a legal/copyright disclaimer for its images. This proposal uses only its `prompt`
field and does not download, train on, or redistribute its image or LLaVA fields.

#### UniCoT direct parsing

For UniCoT state `i`, the source adapter maps:

```text
current image = input_image[i]
reflection    = eval_summary[i], or cleaned/length-limited eval[i]
action        = continue iff output_image[i] is non-null; otherwise stop
edit          = edit[i] for continue; empty for stop
next image    = output_image[i] for continue
```

`Action` is derived from the transition structure rather than from natural-language phrases. Every non-terminal
`output_image[i]` must hash-match `input_image[i + 1]`.

Length mismatches, empty reflections, `continue` with an empty Edit, contradictory terminal rows, and missing images fail
closed.

#### Echo-4o 0/1-turn synthesis

For an Echo T2I pair `(prompt p, reference image I_ref)`:

```text
I_0 = repo-native BAGEL.generate(p, stable_seed)

reflector VLM(p, I_0)
  -> Reflection_0 / Action_0 / Edit_0

if Action_0 == stop:
    verifier VLM(p, I_0, proposed response)
      -> accept 0-turn, retry, or reject

if Action_0 == continue:
    verifier VLM(p, I_0, I_ref, proposed response)
      -> confirm edit compatibility, retry, or reject

    terminal reflector VLM(p, I_ref)
      -> Reflection_1 / stop / empty Edit

    terminal verifier
      -> accept 1-turn, retry, or reject
```

The accepted trajectories are:

```text
0-turn:
[I_0, Reflection_0 / stop / empty Edit]

1-turn:
[I_0, Reflection_0 / continue / Edit_0,
 I_ref, Reflection_1 / stop / empty Edit]
```

The initial reflector does not see `I_ref`; the reference is available only to the offline verifier. The verifier must not
silently rewrite `Action_0`. A zero-turn sample requires a valid `stop` response, while a one-turn sample requires a valid
`continue` response and non-empty Edit.

`I_ref` is accepted as the edit target only when subject, scene, and composition remain aligned, the proposed Edit
explains the delta, the defect is corrected, and unrelated content is preserved. This route does not call an edit model.

#### Midjourney prompt-driven k-turn synthesis

For each accepted Midjourney prompt `p`:

```text
I_0 = repo-native BAGEL.generate(p, stable_seed)

repeat up to max_edit_turns:
    VLM(p, I_k)
      -> Reflection_k / Action_k / Edit_k

    if Action_k == stop:
        verifier VLM confirms the terminal state
        finish

    I_(k+1) = edit_model(p, I_k, Edit_k)

    verifier VLM checks:
        edit adherence
        prompt fidelity
        quality improvement
        preservation
```

Every accepted edit must be followed by another reflection. Structured parse failure, `continue` with an empty Edit,
edit non-adherence, quality regression, a missing terminal `stop`, or reaching `max_edit_turns` while still requesting
`continue` causes a bounded retry followed by rejection.

The VLM and edit model are offline data-preparation dependencies, not runtime dependencies of BAGEL SFT.

### Canonical trajectory and atomic planner

All three data routes produce the same tokenizer-independent trajectory:

```python
class ImageRef(TypedDict):
    uri: str
    sha256: str


class ReflectionStep(TypedDict):
    reflection: str
    action: Literal["continue", "stop"]
    edit: str


class VisualReflectionTrajectory(TypedDict):
    trajectory_id: str
    manifest_id: str
    source_dataset: str
    source_record_id: str
    pipeline_variant: Literal[
        "direct_parse",
        "pair_0_1_turn",
        "prompt_k_turn",
    ]
    prompt: str
    images: list[ImageRef]
    steps: list[ReflectionStep]
```

For a trajectory containing `N` images:

- `len(images) == len(steps) == N`;
- the first `N - 1` steps are `continue` with non-empty delta Edit instructions;
- every `continue` step has a supervised next image;
- the final step is `stop` with an empty Edit.

The proposed `t2i_target_policy = terminal_state_v1` uses the final trajectory image as the T2I target rather than a
draft already labeled as defective.

Each trajectory deterministically expands into:

```text
1     t2i entry:
      prompt -> image_(N-1)

N     reflect entries:
      prompt + image_i
        -> Reflection_i / Action_i / Edit_i

N - 1 edit entries:
      prompt + image_i + Edit_i
        -> image_(i+1)
```

Train/validation/test splitting and deduplication happen at the source-record or trajectory level before synthesis and
atomic expansion, so entries derived from the same source cannot cross splits.

### Reflection protocol

The MVP uses ordinary structured text supported by the existing tokenizer:

```text
Reflection: The subject is correct, but the left hand is distorted.
Action: continue
Edit: Fix the left hand while preserving the pose and background.
```

Terminal response:

```text
Reflection: The image now follows the prompt with no obvious defect.
Action: stop
Edit:
```

Only lowercase `continue` and `stop` are allowed. Missing or duplicate fields, unknown actions, `continue` with an empty
Edit, and `stop` with a non-empty Edit fail validation.

No special tokens, vocabulary changes, or separate action/Edit losses are proposed. `Reflection` is visible supervised
assessment text, not a hidden reasoning channel.

### Repo-native BAGEL changes

The proposed model structure is:

```text
repo-native BAGEL shared components
  |- BagelForTraining  # existing FlowGRPO compatibility boundary
  `- BagelForSFT       # t2i / reflect / edit SFT boundary
```

`BagelForSFT` should reuse the current repository-owned MoT and generation components while adding or loading:

- the text `lm_head`;
- the ViT, connector, and related position components;
- a frozen VAE encoder for clean-source image conditioning;
- task-specific sequence construction, MoT routing, attention masks, and label masks;
- typed `forward_t2i`, `forward_reflect`, and `forward_edit` methods.

For `reflect`, the clean image and prompt form the context, and only the serialized assistant response receives causal-LM
supervision.

For `edit`, clean-source VAE/ViT representations, the original prompt, and the delta Edit condition the next-image target.
The next image must not appear in the clean conditioning prefix.

The existing `BagelForTraining` API, checkpoint mapping, weight-sync parameter names, trainable-parameter policy, and
fixed-input FlowGRPO output remain compatibility boundaries.

### Losses

The `reflect` primitive uses an understanding loss, `L_und`. Let `R_i,t` select labels belonging to the serialized
`Reflection` / `Action` / `Edit` response and EOS:

```math
L_{\mathrm{und}}
=
-\frac{1}{|B_{\mathrm{und}}|}
\sum_{i\in B_{\mathrm{und}}}
\frac{
\sum_t R_{i,t}\log p_\theta(x_{i,t+1}\mid x_{i,\le t})
}{
\sum_tR_{i,t}
}.
```

Image/context tokens, the original prompt, system text, and padding are excluded from this CE.

For the image primitives, let `M_i,j` denote the valid target-image mask. The T2I loss is:

```math
L_{\mathrm{t2i}}
=
\frac{1}{|B_{\mathrm{t2i}}|}
\sum_{i\in B_{\mathrm{t2i}}}
\frac{
\sum_j M_{i,j}
\left\lVert
\hat v^{\mathrm{t2i}}_{i,j}
-v^{\mathrm{t2i,target}}_{i,j}
\right\rVert_2^2
}{
d_v\sum_jM_{i,j}
}.
```

The edit loss is:

```math
L_{\mathrm{edit}}
=
\frac{1}{|B_{\mathrm{edit}}|}
\sum_{i\in B_{\mathrm{edit}}}
\frac{
\sum_j M_{i,j}
\left\lVert
\hat v^{\mathrm{edit}}_{i,j}
-v^{\mathrm{edit,target}}_{i,j}
\right\rVert_2^2
}{
d_v\sum_jM_{i,j}
}.
```

The current image remains clean condition, and the next image cannot appear in the clean prefix. The total objective is:

```math
L
=
\lambda_{\mathrm{und}}L_{\mathrm{und}}
+\lambda_{\mathrm{t2i}}L_{\mathrm{t2i}}
+\lambda_{\mathrm{edit}}L_{\mathrm{edit}}.
```

Each homogeneous micro-batch executes one primitive. Task-sampling ratios, objective weights, and raw/weighted losses are
explicitly configured and logged.

### Scope boundaries

This RFC does not propose:

- a generic UMM base class;
- whole-trajectory packing;
- online multi-turn RL or reward-model changes;
- tokenizer vocabulary changes;
- a full-weight training recipe;
- a general-purpose teacher/editor serving system inside verl-omni.

The initial implementation target is CUDA + FSDP2 + LoRA.

### Public data/code references

- [UniCoT-Self-Reflection-6K dataset](https://huggingface.co/datasets/Fr0zencr4nE/UniCoT-Self-Reflection-6K)
- [UniCoT code](https://github.com/Fr0zenCrane/UniCoT)
- [Echo-4o-Image dataset](https://huggingface.co/datasets/Yejy53/Echo-4o-Image)
- [Echo-4o code](https://github.com/yejy53/Echo-4o)
- [Midjourney-v6-LLaVA dataset](https://huggingface.co/datasets/brivangl/midjourney-v6-llava)
- [Current verl-omni BAGEL implementation](https://github.com/verl-project/verl-omni/tree/main/verl_omni/pipelines/bagel_flow_grpo)

### Motivation

verl-omni currently supports BAGEL primarily through the FlowGRPO text-to-image training path. That path can optimize generated images using rollout trajectories and visual rewards, but it does not provide an offline SFT contract for teaching BAGEL to:

- inspect a generated image through its visual-understanding path;
- produce a structured visible reflection;
- decide whether to `continue` or `stop`;
- generate a concrete delta Edit instruction;
- condition image editing on the current image; and
- repeat this process for a bounded number of turns.

The current training-facing BAGEL module exposes the generation path required by FlowGRPO, but it does not yet expose the `lm_head`, ViT/connector understanding path, clean-source image conditioning, or an offline SFT engine for the three objectives proposed above.

This feature would allow BAGEL to learn the complete behavior:

```text
generate -> reflection -> edit -> reflection again
```

while preserving the existing FlowGRPO path and reusing verl-omni's model loading, FSDP2, LoRA, checkpoint, worker, and tracking infrastructure.

Related work should be coordinated with:

- [#217](https://github.com/verl-project/verl-omni/issues/217), which may overlap with generic image-data helpers and atomic image SFT;
- [#232](https://github.com/verl-project/verl-omni/issues/232), whose thinker/talker abstraction is not modified by this proposal.

### Your contribution

I can help implement this RFC in four reviewable stages.

### PR 1 — `[data,tests] feat: add BAGEL visual-reflection data contracts and public adapters`

Proposed scope:

- canonical trajectory and reflection-step schemas;
- UniCoT direct parser;
- Echo T2I-pair source adapter;
- Midjourney prompt-only source adapter;
- synthesis input/output and provenance contracts;
- strict formatter/parser and rejection reasons;
- deterministic atomic planner;
- source-level split and deduplication;
- CPU tests for valid and malformed 0/1/k-turn fixtures.

This PR would not load BAGEL weights.

### PR 2 — `[model,tests] feat: add repo-native BAGEL atomic SFT primitives`

Proposed scope:

- refactor reusable BAGEL components while preserving `BagelForTraining`;
- add `BagelForSFT`;
- load the `lm_head`, ViT, connector, and clean-source VAE components;
- implement `forward_t2i`, `forward_reflect`, and `forward_edit`;
- implement `L_und`, `L_t2i`, and `L_edit`;
- add primitive-loss, attention-leakage, checkpoint-key/shape, and FlowGRPO parity tests.

### PR 3 — `[trainer,fsdp,ckpt] feat: add FSDP2 atomic SFT for BAGEL`

Proposed scope:

- BAGEL-specific offline SFT trainer/engine;
- homogeneous task sampling;
- configurable task/loss weights;
- FSDP2 + LoRA training;
- gradient accumulation and distributed normalization;
- task-sampler/dataloader resume state;
- LoRA artifact and merged repo-native BAGEL export;
- one-rank/two-rank, resume, reload, and FlowGRPO initialization tests.

### PR 4 — `[tests,doc] feat: add BAGEL reflection-loop evaluation and data recipes`

Proposed scope:

- teacher-forcing evaluation for all three primitives;
- bounded generate-reflect-edit evaluator;
- direct `stop` and `continue -> edit -> stop` tests;
- malformed-response and maximum-turn handling;
- tiny-set overfit test;
- public data-conversion and synthesis recipes;
- training, resume, export, and evaluation documentation.

The dependency order would be:

```text
PR 1 -> PR 2 -> PR 3 -> PR 4
```

I will follow the repository's [CONTRIBUTING.md](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md) and run the duplicate-work checks required by `AGENTS.md` before starting each PR. Each PR will include the exact test commands/results, explain why it does not duplicate an existing PR, and disclose AI assistance. All generated changes will be reviewed and understood by the human submitter before submission.

**Comment by wtomin** [comment]
It is very nice that you can take `pr1` and `pr2` tasks, and I will take `pr3` task. Let's work together @BorisYang326

**Comment by SamitHuang** [comment]
Interesting.
1. What is the model used for reflection btw?
2. Why "prompt -> generate -> reflect" rather than "prompt -> reasoning, prompt re-writing -> generate"?

**Comment by BorisYang326** [comment]
@SamitHuang Thanks — two clarifications:

1. At SFT training and evaluation time, the reflection model is repo-native BAGEL itself. In PR2 (#333), `BagelForSFT.forward_reflect` uses BAGEL's ViT/connector understanding path and language head to predict the visible structured `Reflection` / `Action` / `Edit` response.

The offline data-construction teacher depends on the source. UniCoT trajectories are parsed directly through the generic contracts in PR1 (#313), while the Echo and Midjourney synthesis routes may use an external VLM to construct and verify supervision. Those teacher VLMs are offline data-preparation dependencies only; they are not runtime dependencies of the trained BAGEL reflection loop.

2. We use `prompt -> generate -> reflect` because this RFC targets post-generation visual self-correction. Many failures — distorted anatomy, missing objects, incorrect spatial relationships, prompt-fidelity failures, or unsuccessful edits — can only be verified against the image that was actually produced. Reflection therefore inspects the realized image, decides `continue` or `stop`, and emits a concrete delta Edit instruction when correction is needed.

`prompt -> reasoning / prompt rewriting -> generate` is complementary, but it cannot verify whether the generated pixels actually satisfy the rewritten prompt. It could be added later as a separate pre-generation primitive; the current MVP deliberately isolates three atomic behaviors: `t2i`, `reflect`, and `edit`.
